### ISSUE TITLE
Multiple commits tagged when using `git` plugin

### ISSUE THREAD

**Issue by effervescentia** [semantic-release/semantic-release#649]
## Current behavior

In a configuration with custom `publish` plugins, a tag is created and pushed for the initial commit hash before the `@semantic-release/github` and `@semantic-release/git` plugins are being used to create and tag a new commit with an updated changelog from `@semantic-release/changelog`
Because of this my CI now gets 2 webhook notifications for new tags being pushed and has to run 2 deployment workflows 

## Expected behavior

When overriding the `publish` plugins nothing else should create or push tags to github

## Environment

- **semantic-release** version: `13.1.3`
- CI environment: Circle CI
- Plugins used: `@semantic-release/changelog`, `@semantic-release/git`, `@semantic-release/github`
- **semantic-release** configuration:
```
"release": {
    "tagFormat": "v${version}-dev",
    "verifyConditions": [
      "@semantic-release/changelog",
      "@semantic-release/git",
      "@semantic-release/github"
    ],
    "publish": [
      {
        "path": "@semantic-release/changelog",
        "changelogFile": "DEV_CHANGELOG.md"
      },
      {
        "path": "@semantic-release/git",
        "message": "chore(dev release): ${nextRelease.version}-dev [skip ci]\n\n${nextRelease.notes}",
        "assets": "DEV_CHANGELOG.md"
      },
      "@semantic-release/github"
    ]
  },
```
- CI logs: (sorry, private CI)
```
[Semantic release]: Load plugin verifyConditions from @semantic-release/git in shareable config ./release.dev
[Semantic release]: Load plugin verifyConditions from @semantic-release/github in shareable config ./release.dev
[Semantic release]: Load plugin analyzeCommits from @semantic-release/commit-analyzer
[Semantic release]: Load plugin generateNotes from @semantic-release/release-notes-generator
[Semantic release]: Load plugin publish from @semantic-release/changelog in shareable config ./release.dev
[Semantic release]: Load plugin publish from @semantic-release/git in shareable config ./release.dev
[Semantic release]: Load plugin publish from @semantic-release/github in shareable config ./release.dev
[Semantic release]: Run automated release from branch master
[Semantic release]: Call plugin verify-conditions
[Semantic release]: Verify GitHub authentication
[Semantic release]: Found git tag v0.2.1-dev associated with version 0.2.1
[Semantic release]: Found 1 commits since last release
[Semantic release]: Call plugin analyze-commits
[Semantic release]: Analyzing commit: feat(CC-473): form styling for remaining models (#100)
[Semantic release]: The release type for the commit is minor
[Semantic release]: Analysis of 1 commits complete: minor release
[Semantic release]: The next release version is 0.3.0
[Semantic release]: Call plugin verify-release
[Semantic release]: Call plugin generateNotes
[Semantic release]: Create tag v0.3.0-dev
[Semantic release]: Call plugin publish
[Semantic release]: Update DEV_CHANGELOG.md
[Semantic release]: Found 1 file(s) to commit
[Semantic release]: Creating tag v0.3.0-dev
[Semantic release]: Published Git release: v0.3.0-dev
[Semantic release]: Create tag v0.3.0-dev
[Semantic release]: Call plugin generateNotes
[Semantic release]: Published GitHub release: https://github.com/groupby/command-center/releases/tag/v0.3.0-dev
[Semantic release]: Published release: 0.3.0
```

**Comment by pvdlg** [comment]
semantic-release create the tag before calling the `publish` plugins and update the tag reference if a `publish` adds a commit.
We do that because certain `publish` plugin require to have the tag created on the remote repository before running. This is the case for Atom packages as `apm publish` require to have the tag created on the remote repo.

Unfortunately, if you use the `@sematnic-release/git` to create a commit that result in 2 tag creation events.

I don't know how both requirements can be achieved unfortunately. Do you have any suggestions?

In you case you could use something like [Travis auto cancellation](https://blog.travis-ci.com/2017-03-22-introducing-auto-cancellation) as a workaround. This way the job triggered by the first even would be canceled by the second one.

**Comment by effervescentia** [comment]
What if instead of creating it before `publish` it becomes one of the plugins that runs in the `publish` pipeline?
That way if someone is doing a custom release to `apm` then they would just need to include the `@semantic-release/git-tag` plugin in their `publish` array before their `apm publish` script

It could even become part of the functionality of `@semantic-release/git` for simplicity and a configuration flag could be added to indicate whether or not `@semantic-release/git` should do commit in order to avoid the case when it would error on committing no changes. Alternatively it could fail silently on a "no changes to commit" error.

Unfortunately my CI doesn't have auto-cancellation currently 😞

**Comment by pvdlg** [comment]
Well, that's the architecture we moved away from recently: #613. There is several benefits to handling the git tag in the core, mostly that the creation is mandatory in order to identify the last release, so we can analyze commits between then and now.

Also we can't presume that a user will use a plugin or not. So all other plugin were creating or re-creating their own tags, just in case a previous plugin didn't do it already.
Basically plugins are meant to be independents and optional. What you suggest would make `@semantic-release/git` mandatory and would make every other plugin depending on it (for example you would need to run `@semantic-release/git` first so `@semantic-release/github` could create a release on the tag created by `@semantic-release/git`).

The current architecture (with the core creating the tag) is much more simple, and make plugin creation a lot easier as they don't have to worry about checking if a tag exist, creating it, re-create it if the added a commit or all the authentication issue that come with it.

So that's a tradeoff. And to be honest we didn't think they would be any inconvenient moving the tag creation to the core as we didn't think about the double tag creation event.

May I ask what you are doing in the CI job triggered by the tag event?
Maybe you could do that as part of `sematnic-release` with the [exec plugin](https://github.com/semantic-release/exec). You could add a last `publish` step that runs whatever you are currently running in the job triggered by the tag event.
That would also have the advantage of doing everything in the same job, avoiding to wait for a new one to start on the tag event.

**Comment by effervescentia** [comment]
The tag event triggers a pipeline of other jobs that culminate in pushing a finalized docker container, and ultimately will be hooked up to our deployment process, all of which unfortunately is more complicated than I would like for the `exec` plugin to be responsible for.

I understand the benefits of having the tag creation happen in the core, and given the insight you've just shared I agree it shouldn't move. Is it possible to make it conditional as to whether or not the tag created in core is actually pushed to github? I think you should be able to do the same commit analysis using the local repository. So otherwise I believe the functionality would remain the same.

If that isn't possible, what if you were to add a pipeline of plugins that would execute before the tag gets created?

So the lifecycle would change from:
- `verify` (etc.)
- `create tag & push tag`
- `publish`
- `done`

to 

- `verify` (etc.)
- `create tag`
-  `pre-push` <- this is where I could add plugins that add any new commits, as long as a tag by the same name as the one created in `create tag` exists before `push tag`
- `push tag`
- `publish`
- `done`

**Comment by pvdlg** [comment]
> Is it possible to make it conditional as to whether or not the tag created in core is actually pushed to github?

Unfortunately `apm` requires to have the tag on the remote. When calling `apm publish` it will make a call to the Atom registry, that will in turn make a call to GitHub API to retrieve the code flagged by the tag.

> If that isn't possible, what if you were to add a pipeline of plugins that would execute before the tag gets pushed?

I think that would be possible. Wouldn't the steps be as follow instead?
1. verify
2. pre-push
3. create tag (local and remote)
4. publish

In pre-push you can have plugins that created commits, while in publish they couldn't.

**Comment by pvdlg** [comment]
Here is a proposal to add a `prepare` plugin hook.

### semantic-release changes

- Add the plugin `prepare` hook that would work similarly to the current `publish`
- Create the tag after `prepare` and before `publish`
- Remove the now unnecessary logic to update the tag if a plugin had created a commit
- Add `@semantic-release/npm` to the default `prepare` hook

### @semantic-release/npm changes
- Move the `pacakge.json` version update and `npm pack` to a new `prepare` hook
- Remove the `npmPublish` option as it would be useless now (user can just disable the publish part to not release on npm)

### @semantic-release/changelog changes
- Change the `publish` step to `prepare`

### @semantic-release/git changes
- Change the `publish` hook to `prepare`

### @semantic-release/exec changes
- Add the `prepare` hook

### @semantic-release/apm-config changes
- Move @semantic-release/git from `publish` to `prepare`

@effervescentia, @gr2m what do you think? Would that work?

**Comment by effervescentia** [comment]
@pvdlg I think that looks awesome!
I believe had only suggested my initial flow in case there were any requirements on the tag already being created for the commit analysis. Basically doesn't matter as long as all of the parameters that are used by the current `publish` plugins are available in `prepare` then it does everything I could ask ever for 👍 💯 
Thanks for being so expedient 😄

**Comment by pvdlg** [comment]
Thanks for suggesting it! I was not really happy with the trick that consisted in updating the tag when a plugin was making a commit anyway.
Splitting the `prepare` and `publish` would also open more opportunities in the future such splitting the release in multiple independent steps (e.g ones could execute semantic-release with some plugins in `prepare` but none in `publish`, and would do their own publish stuff in a job triggered by the tag creation).

**Comment by effervescentia** [comment]
I was thinking the same thing, makes it even more flexible to use.

PS I wanted to say I'm really grateful for all of the work that went into the rewrite. I've been using `semantic-release` for over a year now and I couldn't imagine development without it. The new configuration makes it a lot more customize-able, while I know the old version often confused my coworkers as being a bit of a black box. I actually wrote a plugin ([start-release](https://www.npmjs.com/package/start-release)) for a tiny build framework `start` and it was a little prickly to pick through the source code and wrap the package itself.
I'm not sure how you are for contributors, but I'd be happy to help implement this change if you'd be interested.

Thanks again for taking this feature request 😄

**Comment by pvdlg** [comment]
Contributions are always welcome! In that case I don't think it's the most trivial change, especially for the core, and because it involve changes across multi-repo.
I'd invite to look around in the code and see which of the changes mentioned you'd like to work on (it can be plugin for example or the whole thing if you feel confident). Then we can split the work if you'd like.

But before making any change on this we need to have everybody in the maintainers team on board with it :)

More generally we flag all the issues ready to worked on with the [help wanted tag](https://github.com/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+user%3Asemantic-release+archived%3Afalse+label%3A%22help+wanted%22). So we need plenty of help for those!

**Comment by pvdlg** [comment]
Fixed by #663.

**Comment by effervescentia** [comment]
@pvdlg wanted to let you know I set this up today and it works perfectly 😄  thanks again
